Reuse Lazy IR nodes for faster tracing

desertfire opened the pull request:
Summary: Use a trie to store IR nodes from previous tracings, and the trie serves as a cache for looking up IR nodes and reuse them if possible.

wconstab commented on lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:
```diff
@@ -675,7 +675,7 @@ void InitLtcModuleBindings(py::module m) {
           for (auto& tensor : tensors) {
             auto xtensor = TryGetLtcTensor(tensor);
-            roots.push_back(xtensor->GetIrValue().node.get());
+            roots.push_back(xtensor->GetIrValue().node().get());
```
what is this for?

wconstab commented on torch/csrc/lazy/ts_backend/ops/device_data.cpp:
```diff
@@ -12,7 +12,7 @@ DeviceData::DeviceData(std::shared_ptr<BackendData> data)
           data->shape(),
           /*num_outputs=*/1,
-          /*hash_seed=*/static_cast<uint32_t>(101)),
+          /*hash_seed=*/Node::NextNodeListIndex()),
```
hmm.. we're saying we want the DeviceData node hash to be sensitive to whether it was used earlier or later in the trace, in addition to tensor shape/dtype.

i think it is correct, but i'm not sure if it helps reuse?

wconstab commented on torch/csrc/lazy/core/ir.cpp:
```diff
@@ -46,21 +51,37 @@ bool Node::enableDynamicShape() {
 }
 
+size_t Node::NextNodeListIndex() {
+  // Tracing is done in a single thread, so no need to use atomic here.
```
i think this is good enough for prototyping, but at some point we'd need a design that is thread safe in case someone did use multiple threads..

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -161,19 +178,42 @@ class TORCH_API Node {
+  } else {
+    TORCH_LAZY_COUNTER("IrNodeReused",1);
+  }
+  Node::PushIntoNodeList(node);
```
Are you sure if all the places that you pass a node around are avoiding to invoke the node's copy constructor?

wconstab commented on torch/csrc/lazy/core/lazy_graph_executor.cpp:
```diff
@@ -475,6 +475,12 @@ void LazyGraphExecutor::MarkStep(const BackendDevice& device) {
+
+  auto tensors = GetLiveTensors(&device);
+  for (LazyTensorPtr& tensor : tensors) {
+    tensor->ResetIrValue();
```
can you explain the rationale for this ResetIrValue here?

wconstab commented:
@desertfire a while back you did some profiling of tracing, you showed a breakdown of the time spent inside one 'LazyNativeFunction' and if my memory serves, the majority of time was spent in node constructor+hash computation, also shape inference.

Are you able to replicate those past results and see if the current code improves them?

Note: I think your current code does not help shape inference, which is still significant

desertfire committed:
Reuse Lazy IR nodes for faster tracing

Summary: Use a trie to store IR nodes from previous tracings, and the
trie serves as a cache for looking up IR nodes and reuse them if
possible.

wconstab commented:
@desertfire what's the status of this- ready for a review, or still working on some things?

desertfire commented:
> @desertfire what's the status of this- ready for a review, or still working on some things?

It is ready to be reviewed, although I still need to port this to master after all the major IR changes are settled down there.

wconstab commented on lazy_tensor_core/lazy_tensor_core/csrc/tensor_aten_ops.cpp:
```diff
@@ -272,14 +279,15 @@ torch::lazy::LazyTensorPtr squeeze(const torch::lazy::LazyTensorPtr& input, int6
 void squeeze_(torch::lazy::LazyTensorPtr& input) {
-  input->SetIrValue(
-      torch::lazy::MakeNode<ir::ops::Squeeze>(input->GetIrValue(), -1));
+  input->SetIrValue(torch::lazy::ReuseOrMakeNode<ir::ops::Squeeze>(
```
its up to you, but i have been deleting the `torch::lazy` prefix in a lot of code once it finally landed in a place where everything was in that namespace.  If you are s/replacing anyway, might be worth trimming it but no big deal to keep it

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -241,18 +255,32 @@ def this_shape(i: int) -> str:
-        TORCH_INTERNAL_ASSERT(shapes.size() == {returns_length});"""
+            TORCH_INTERNAL_ASSERT(shapes.size() == {returns_length});"""
+
+        symbole_str = aten_symbol(schema)
```
symbol (typo?)

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -241,18 +255,32 @@ def this_shape(i: int) -> str:
+        symbole_str = aten_symbol(schema)
+        # Rewrite inplace symbol like relu_ into relu
+        # TODO: zero_ is an exception. Need figure out why.
+        if symbole_str == "at::aten::relu_":
```
mm, we generally support 'functional' versions of ops directly (see ts_ yaml) - we get inplace versions for free via codegenerated helpers that make e.g. foo_(self,  ...) do something like (self = foo()) using copy.  I am not sure why we support `relu_` operator directly.

https://github.com/pytorch/pytorch/commit/57a60c5a12b80f2173d67901a36b47156fde7023
This was the PR that added relu_ - @alanwaketan do you recall why you didn't just add `relu` here?

bdhirsh replied:
It sounds reasonable to me to get rid of `relu_()` and just lower `relu()` directly.

I dunno about `relu` specifically, but historically I think part of the reason there were/are some inplace/out= lowerings in XLA is because either the functional variant of the op didn't exist (or the functional variant was composite and defined in terms of the out= variant, although the situation there is better with structured kernels).

I also see a comment about `zero()` - I actually just added a functional `zero()` op, to be used by the functionalization pass. LTC/XLA will also probably want to use it too (main advantage is probably just that `zero` will be more amenable to the LTC codegen) - https://github.com/pytorch/pytorch/pull/75913

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -241,18 +255,32 @@ def this_shape(i: int) -> str:
+        # Rewrite inplace symbol like relu_ into relu
+        # TODO: zero_ is an exception. Need figure out why.
+        if symbole_str == "at::aten::relu_":
+            symbole_str = symbole_str[:-1]
```
hmm. do we actually want to truncate the symbol? let's assume that for some reason we actually want to support `relu_` instead of relying on the copy trick above.  then why not use its actual _ symbol?

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -241,18 +255,32 @@ def this_shape(i: int) -> str:
+            symbole_str = symbole_str[:-1]
+
+        node_str = f"""torch::lazy::NodePtr node = torch::lazy::ReuseNode<ir::ops::{schema.node_name}>(
+            torch::lazy::OpKind({symbole_str}),
```
i was wondering; is it possible inside ReuseNode<T> to call `OpKind(T)` directly using the template arg? (so you can avoid passing in symbol here?)

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -241,18 +255,32 @@ def this_shape(i: int) -> str:
+            {node_ctor_input_str});
+        if (!node) {{{meta_str}
+            node = torch::lazy::MakeNode<ir::ops::{schema.node_name}>(
+                    {node_ctor_input_str.replace("      ", "          ")},
```
oh jeez

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -273,11 +301,11 @@ def this_shape(i: int) -> str:
 
         return [f"""\
     {sig.decl(name=f"{self.class_method_name}::{metadata.kernel}")} {{
+        TORCH_LAZY_TIMED("LazyNativeFunctions::full_codegen");
```
see my latest refactor of codegen (on master) where more stuff is 'customizable' being string variables in the GenLazyNativeFunc class.  I don't think XLA wants to customize this as they may not even care about it, but cc @JackCaoG in case he does. We probably want to make this a string variable too

wconstab commented on tools/codegen/dest/lazy_ts_lowering.py:
```diff
@@ -39,3 +39,26 @@ def ts_lowering_body(f: Union[NativeFunctionsGroup, NativeFunction]) -> str:
+    for arg in schema.positional_args:
+        if arg.is_lazy_value:
+            if isinstance(arg.lazy_type, OptionalCType):
+                value_comparsion.append(f"(!has_{arg.name} || operand(i++) == {arg.name}.value())")
```
don't we need to check that our has_argname matches their has_argname?  e.g. if we !have_argname, but they have_argname, it is bad to reuse?

wconstab commented on torch/csrc/lazy/core/ir.cpp:
```diff
@@ -21,6 +21,10 @@ std::string Output::ToString() const {
 }
 
+bool Output::operator==(const Value& rhs) const {
+  return node->hash() == rhs.node->hash() && index == rhs.index;
```
hmm. when is Output::operator== used?

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -138,6 +133,22 @@ class TORCH_API Node {
+  // for the operation. The num_outputs tells how many outputs a given operation
+  // generates.
+  //
+  // None leaf node's node_hash does not contains shape information always.
```
`None leaf node's` seems like a typo?

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -138,6 +133,22 @@ class TORCH_API Node {
 
   virtual std::string ToString() const;
 
+ protected:
```
what's the significance of this change to protected?  since we still default to not using 'reuse' mode, i guess it doesn't change any actual behavior (e.g. prohibiting creating nodes).

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -188,6 +193,43 @@ const T* NodeCast(const Node* node, OpKind op) {
+      NodePtr ir_node = (*it)->ir_node;
+      const T* concrete_node = NodeCast<T>(ir_node.get(), op);
+      if (concrete_node && concrete_node->Equal(std::forward<Args>(args)...)) {
+        TORCH_LAZY_COUNTER("IrNodeReused::" + std::string(typeid(T).name()), 1);
```
did you ever count the non-reused nodes and make sure it's not hitting when not expected?

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -188,6 +193,43 @@ const T* NodeCast(const Node* node, OpKind op) {
+}
+
+template <typename T, typename... Args>
+NodePtr ReuseOrMakeNode(OpKind op, Args&&... args) {
```
any reason to not call this version from the codegen?

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -188,6 +193,43 @@ const T* NodeCast(const Node* node, OpKind op) {
+  return nullptr;
+}
+
+// TODO(alanwaketan): Support r-value reference argument type.
```
@alanwaketan is this still worth fixing? is it hard to fix? performance of MakeNode is a little less critical if we land this PR but it would still be nice to do the 'best practices' thing if it's not too difficult

wconstab commented on torch/csrc/lazy/core/lazy_graph_executor.cpp:
```diff
@@ -297,8 +298,14 @@ class DeviceContextArena {
-    Value k = MakeNode<Scalar>(MakeIntScalar(kSeedMul), kSeedType);
-    Value b = MakeNode<Scalar>(MakeIntScalar(kSeedAdd), kSeedType);
+    Value k = ReuseOrMakeNode<Scalar>(
+        torch::lazy::OpKind(at::prim::Constant),
```
hmm. i see now why OpKind(T) could not work.. or at least not all the time

wconstab commented on torch/csrc/lazy/core/lazy_graph_executor.cpp:
```diff
@@ -475,6 +482,8 @@ void LazyGraphExecutor::MarkStep(const BackendDevice& device) {
   DeviceContextArena::Get()->MarkStep(device);
   ScopePusher::ResetScopes();
   g_tls_data.Reset();
+  Trie::Get()->ResetCurrent();
```
I guess it would be nicer to name the class such that when you call this function it's more apparent what the thing is used for.  LazyTraceCache?

wconstab reviewed (commented):
I think when you make PRs to master, you should split this into at least 2 PRs (could be more)
 1) changes to add new infra that is somewhat isolated (along with their test if applicable)
 ...) 
  2) updating all the ops to call ReuseNode
  
  bc it's a little crazy to review so many LOC at once

wconstab commented on torch/csrc/lazy/core/trie.h:
```diff
@@ -0,0 +1,49 @@
+  explicit TrieNode(NodePtr node) : unique_id(GetNextUniqueId()), ir_node(std::move(node)) { }
+};
+
+class TORCH_API Trie {
```
you should add a docstring to the overall Trie (TraceCache?) class.  It should explain basically how it's used.

wconstab commented on torch/csrc/lazy/core/trie.cpp:
```diff
@@ -0,0 +1,74 @@
+}
+
+Trie* Trie::Get() {
+  static Trie* trie = new Trie();
```
static or thread_local?

JackCaoG commented:
> I think when you make PRs to master, you should split this into at least 2 PRs (could be more)
> 
> 1. changes to add new infra that is somewhat isolated (along with their test if applicable)
>    ...)
> 2. updating all the ops to call ReuseNode
> 
> bc it's a little crazy to review so many LOC at once

+1 I think we should break this pr in to smaller peices for review and landing. It is easier to review and also a lot easier to revert if things goes wrong.

wconstab commented on torch/csrc/lazy/core/trie.cpp:
```diff
@@ -0,0 +1,74 @@
+void Trie::Insert(NodePtr ir_node) {
+  TORCH_CHECK(current_);
+  if (!current_->successors.empty()) {
+    TORCH_LAZY_COUNTER("ForkTrie", 1);
```
this is an interesting thing to track; but on its own it isn't that useful bc it doesn't tell us if we have width 2 at 10 layers or width 20 at the first layer.  did you do any instrumentation on torchbench to see how things looked?

wconstab commented on torch/csrc/lazy/core/ir.h:
```diff
@@ -188,6 +193,43 @@ const T* NodeCast(const Node* node, OpKind op) {
+    std::deque<std::unique_ptr<TrieNode>>& successors =
+        Trie::Get()->Current()->successors;
+    for (auto it = successors.begin(); it != successors.end(); it++) {
+      NodePtr ir_node = (*it)->ir_node;
```
would it be easy to move this logic inside of a Trie:: class method (e.g. LookupNode) which accepts std::forward args, and internally handles the SetCurrent logic? I felt like exposing 'SetCurrent' seemed a little odd

wconstab reviewed (commented):
OK i did a full pass. I think overall this looks great. I am wondering how to land/enable it.
1- criteria to enable by default
2- current test coverage strategy
3- compatibility with other backends

What stands in the way of (1)?  We should be running all our test suites with it enabled, and torchbench
